prov/efa/src/efa.h

```diff
@@ -0,0 +1,410 @@
+	uint32_t	desc_offset;
+	size_t		desc_ring_mmap_size;
+	size_t		max_inline_data;
+	size_t		immediate_data_width;
+	uint16_t	sub_cq_idx;
+};
+
+struct efa_rq {
+	struct efa_wq	wq;
+	uint32_t	*db;
+	uint8_t		*buf;
+	size_t		buf_size;
+	uint16_t	sub_cq_idx;
+};
+
+struct efa_qp {
+	struct ibv_qp	ibv_qp;
+	struct efa_ep	*ep;
+	struct efa_sq	sq;
+	struct efa_rq	rq;
+	uint32_t	qp_num;
+	int		page_size;
+};
+
+struct efa_ah {
+	struct ibv_ah	ibv_ah;
+	uint16_t	efa_address_handle;
+};
+
+struct efa_mem_desc {
+	struct fid_mr		mr_fid;
+	struct ibv_mr		*mr;
+	struct efa_domain	*domain;
+	/* Used only in MR cache */
+	struct ofi_mr_entry	*entry;
+};
+
+struct efa_ep {
+	struct fid_ep		ep_fid;
+	struct efa_domain	*domain;
+	struct efa_qp		*qp;
+	struct efa_cq		*rcq;
+	struct efa_cq		*scq;
+	struct efa_av		*av;
+	struct fi_info		*info;
+	void			*src_addr;
+};
+
+typedef struct efa_conn *
+	(*efa_addr_to_conn_func)
+	(struct efa_av *av, fi_addr_t addr);
+
+struct efa_av {
+	struct fid_av		av_fid;
+	struct efa_domain	*domain;
+	struct efa_ep		*ep;
+	size_t			count;
+	size_t			used;
+	size_t			next;
+	uint64_t		flags;
+	enum fi_av_type		type;
+	efa_addr_to_conn_func	addr_to_conn;
+	struct efa_reverse_av	*reverse_av;
+	/* Used only for FI_AV_TABLE */
+	struct efa_conn **conn_table;
+};
+
+struct efa_ah_qpn {
+	uint16_t efa_ah;
+	uint16_t qpn;
+};
+
+struct efa_reverse_av {
+	struct efa_ah_qpn key;
+	fi_addr_t fi_addr;
+	UT_hash_handle hh;
+};
+
+struct efa_ep_domain {
+	char		*suffix;
+	enum fi_ep_type	type;
+	uint64_t	caps;
+};
+
+struct efa_device_attr {
+	struct ibv_device_attr	ibv_attr;
+	uint32_t		max_sq_wr;
+	uint32_t		max_rq_wr;
+	uint16_t		max_sq_sge;
+	uint16_t		max_rq_sge;
+};
+
+static inline struct efa_device *to_efa_dev(struct ibv_device *ibdev)
+{
+	return container_of(ibdev, struct efa_device, verbs_dev);
+}
+
+static inline struct efa_context *to_efa_ctx(struct ibv_context *ibctx)
+{
+	return container_of(ibctx, struct efa_context, ibv_ctx);
+}
+
+static inline struct efa_pd *to_efa_pd(struct ibv_pd *ibpd)
+{
+	return container_of(ibpd, struct efa_pd, ibv_pd);
+}
+
+static inline struct efa_cq *to_efa_cq(struct ibv_cq *ibcq)
+{
+	return container_of(ibcq, struct efa_cq, ibv_cq);
+}
+
+static inline struct efa_qp *to_efa_qp(struct ibv_qp *ibqp)
+{
+	return container_of(ibqp, struct efa_qp, ibv_qp);
+}
+
+static inline struct efa_ah *to_efa_ah(struct ibv_ah *ibah)
+{
+	return container_of(ibah, struct efa_ah, ibv_ah);
+}
+
+static inline unsigned long align(unsigned long val, unsigned long align)
+{
+	return (val + align - 1) & ~(align - 1);
+}
+
+static inline uint32_t align_up_queue_size(uint32_t req)
+{
+	req--;
+	req |= req >> 1;
+	req |= req >> 2;
+	req |= req >> 4;
+	req |= req >> 8;
+	req |= req >> 16;
+	req++;
+	return req;
+}
+
+#define is_power_of_2(x) (!(x == 0) && !(x & (x - 1)))
+#define align_down_to_power_of_2(x)		\
+	({					\
+		__typeof__(x) n = (x);		\
+		while (n & (n - 1))		\
+			n = n & (n - 1);	\
+		n;				\
+	})
```

Comment:
For the above functions/macros, see ofi.h: roundup_power_of_two, fi_get_aligned_sz.  The new calls can be moved there for general use.

Reply:
Good point, will move it up to the common header.